internal/manager/controllers_test.go

```diff
@@ -0,0 +1,186 @@
+package manager
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/manager/index"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/manager/managerfakes"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/manager/predicate"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/reconciler"
+)
+
+func TestCreateFilterForGatewayClass(t *testing.T) {
+	const gcName = "my-gc"
+
+	filter := createFilterForGatewayClass(gcName)
+	if filter == nil {
+		t.Fatal("createFilterForGatewayClass() returned nil")
+	}
+
+	tests := []struct {
+		nsname   types.NamespacedName
+		expected bool
+	}{
+		{
+			nsname:   types.NamespacedName{Name: gcName},
+			expected: true,
+		},
+		{
+			nsname:   types.NamespacedName{Name: gcName, Namespace: "doesn't matter"},
+			expected: true,
+		},
+		{
+			nsname:   types.NamespacedName{Name: "some-gc"},
+			expected: false,
+		},
+	}
+
+	for _, test := range tests {
+		result, msg := filter(test.nsname)
+
+		if result != test.expected {
+			t.Errorf("filter(%#v) returned %v but expected %v", test.nsname, result, test.expected)
+		}
+
+		if result && msg != "" {
+			t.Errorf("filter(%#v) returned a non-empty message %q", test.nsname, msg)
+		}
+		if !result && msg == "" {
+			t.Errorf("filter(%#v) returned an empty message", test.nsname)
+		}
+	}
+}
+
+func TestRegisterController(t *testing.T) {
+	defer func() {
+		newReconciler = reconciler.NewImplementation
```

Comment:
Can you add a comment here about the purpose of this defer? If this was accidentally removed, could it disrupt later tests?

Reply:
Right now it cannot. But we might have a test in the future that could depend on it, so better to restore to the original - that was my thinking. I slightly change the defer to save it in a temp variable.